src/symbol.c

```diff
@@ -0,0 +1,167 @@
+#define R_NO_REMAP
+#include <R.h>
+#include <Rinternals.h>
+#include <R_ext/GraphicsEngine.h>
+#include <stdbool.h>
+
+#define attribute_hidden
+
+// Interface functions ---------------------------------------------------------
+
+void copy_character(SEXP tgt, SEXP src, R_xlen_t len);
+R_xlen_t unescape_character_in_copy(SEXP tgt, SEXP src, R_xlen_t i);
+SEXP unescape_sexp(SEXP name);
+
+SEXP rlang_symbol(SEXP chr) {
+  SEXP string = STRING_ELT(chr, 0);
+  return Rf_install(Rf_translateChar(string));
+}
+
+SEXP rlang_symbol_to_character(SEXP chr) {
+  SEXP name = PRINTNAME(chr);
+  return Rf_ScalarString(unescape_sexp(name));
+}
+
+SEXP rlang_unescape_character(SEXP chr) {
+  R_xlen_t len = Rf_xlength(chr);
+  R_xlen_t i = unescape_character_in_copy(R_NilValue, chr, 0);
+  if (i == len) return chr;
+
+  SEXP ret = PROTECT(Rf_allocVector(STRSXP, len));
+  copy_character(ret, chr, i);
+  unescape_character_in_copy(ret, chr, i);
+  UNPROTECT(1);
+  return ret;
+}
+
+// Private functions -----------------------------------------------------------
+
+bool has_unicode_escape(const char* chr);
+int unescape_char(char* chr);
+int unescape_char_found(char* chr);
+int process_byte(char* tgt, char* const src, int* len_processed);
+bool has_codepoint(const char* src);
+bool is_hex(const char chr);
+
+void copy_character(SEXP tgt, SEXP src, R_xlen_t len) {
+  for (int i = 0; i < len; ++i) {
+    SET_STRING_ELT(tgt, i, STRING_ELT(src, i));
+  }
+}
+
+R_xlen_t attribute_hidden unescape_character_in_copy(SEXP tgt, SEXP src, R_xlen_t i) {
+  R_xlen_t len = Rf_length(src);
+  int dry_run = Rf_isNull(tgt);
+
+  for (; i < len; ++i) {
+    SEXP old_elt = STRING_ELT(src, i);
+    SEXP new_elt = unescape_sexp(old_elt);
+    if (dry_run) {
+      if (old_elt != new_elt) return i;
+    }
+    else {
+      SET_STRING_ELT(tgt, i, new_elt);
+    }
+  }
+
+  return i;
+}
+
+SEXP attribute_hidden unescape_sexp(SEXP name) {
+  int ce = Rf_getCharCE(name);
+  const char* src = CHAR(name);
+  const char* re_enc = Rf_reEnc(src, ce, CE_UTF8, 0);
+
+  char* tmp;
+  if (re_enc != src) {
+    // If the string has been copied, it's safe to use as buffer
+    tmp = (char*)re_enc;
+  }
+  else {
+    // If not, we're in a UTF-8 locale
+    // Need to check first if the string has any UTF-8 escapes
+    if (!has_unicode_escape(src)) return name;
+    int orig_len = strlen(re_enc);
+    tmp = alloca(orig_len + 1);
```

Comment:
I don't understand this remark. Which other allocator are you suggesting to use?